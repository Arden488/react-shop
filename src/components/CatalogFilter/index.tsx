import * as React from 'react';

const CatalogFilter = () => (
  <div>
    <div>
      <div>Product series:</div>
      <div>
        <input type="checkbox" value="" name="" /> iPhone X<br />
        <input type="checkbox" value="" name="" /> iPhone 8<br />
        <input type="checkbox" value="" name="" /> iPhone 7<br />
        <input type="checkbox" value="" name="" /> iPhone 6<br />
      </div>
    </div>
    <div>
      <div>Color:</div>
      <div>
        <input type="checkbox" value="" name="" /> Gray space<br />
        <input type="checkbox" value="" name="" /> Black<br />
        <input type="checkbox" value="" name="" /> Gold<br />
        <input type="checkbox" value="" name="" /> Gold rose
      </div>
    </div>
    <div>
      <div>Price range:</div>
      <div>From 0 to $1000</div>
    </div>
    <div>
      <div>Screen size:</div>
      <div>
        <input type="checkbox" value="" name="" /> 4'<br />
        <input type="checkbox" value="" name="" /> 5'<br />
        <input type="checkbox" value="" name="" /> 6'
      </div>
    </div>
    <div>
      <div>Memory size:</div>
      <div>
        <input type="checkbox" value="" name="" /> 16 Gb<br />
        <input type="checkbox" value="" name="" /> 32 Gb<br />
        <input type="checkbox" value="" name="" /> 64 Gb<br />
        <input type="checkbox" value="" name="" /> 128 Gb<br />
        <input type="checkbox" value="" name="" /> 256 Gb
      </div>
    </div>
    <div>
      <div>Release year:</div>
      <div>
        <div>
          <input type="checkbox" value="" name="" /> 2012<br />
          <input type="checkbox" value="" name="" /> 2011<br />
          <input type="checkbox" value="" name="" /> 2010<br />
          <input type="checkbox" value="" name="" /> 2009<br />
          <input type="checkbox" value="" name="" /> 2008
        </div>
        <a href="#">more</a>
      </div>
    </div>
  </div>
);

export default CatalogFilter;
