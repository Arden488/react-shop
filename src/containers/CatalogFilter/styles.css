.catalog-filter__section {
  margin-bottom: 20px;
}
