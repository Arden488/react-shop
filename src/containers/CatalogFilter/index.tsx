import * as React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { fetchFiltersByCategory, passFilters } from '../../actions';

import './styles.css';

interface ICatalogFilterProps {
  match: any,
  history: any,
  products: any,
  filters: any,
  passFilters: any,
  fetchFiltersByCategory: any,
}

interface ICatalogFilterState {
  filterProps: object,
  priceFrom: string,
  priceTo: string,
}

// tslint:disable-next-line
interface CatalogFilter {
  category: string,
  options: {
    priceTo?: string,
    priceFrom?: string,
  },
}

class CatalogFilter extends React.Component<ICatalogFilterProps, ICatalogFilterState> {
  constructor(props: any) {
    super(props);

    this.options = {};
    this.handlePropertyPriceRange = this.handlePropertyPriceRange.bind(this);
    this.handlePropertyCheck = this.handlePropertyCheck.bind(this);

    this.state = {
      filterProps: {},
      priceFrom: '0',
      priceTo: '1000',
    }

    this.category = this.props.match.params.cat;
  }

  public render() {
    const filters = this.props.filters;
    const filterSections = this.outputFilters(filters);
    
    return (
      <div>
        {filterSections}
      </div>
    )
  }

  public componentDidMount() {
    this.props.fetchFiltersByCategory(this.category);

    this.props.history.listen(() => {
      const path = this.props.history.location.pathname;
      this.category = path.split('/')[2];
      this.props.fetchFiltersByCategory(this.category);
      this.props.passFilters({});
    });
  }

  private getFilterFields(
      title: string,
      name: string,
      fieldType: string, 
      variants: [] = [], 
      options: { min: string, max: string } = { max: '', min: '' }
    ) {
    switch (fieldType) {
      case 'checkbox':
        return this.outputCheckboxes(name, variants);
      case 'range':
        return this.outputRange(name, options);
      default:
        return '';
    }
  }

  private getFilterSection(filter: { 
      title: string,
      name: string, 
      field_type: string, 
      variants: [], 
      options: { min: string, max: string } 
    }) {
    const fields = this.getFilterFields(
      filter.title,
      filter.name, 
      filter.field_type, 
      filter.variants, 
      filter.options
    );

    return (
      <div className="catalog-filter__section" key={filter.name}>
        <div>{filter.title}:</div>
        <div>
          {fields}
        </div>
      </div>
    )
  }

  private outputCheckboxes(name: string, variants: []) {
    const checkboxes = variants.map((item: string) => {
      return (
        <div key={item}>
          <input 
            type="checkbox" 
            value={item} 
            onChange={this.handlePropertyCheck} 
            className="products-filter-property" 
            name={name} /> 
          {item}
        </div>
      )
    });

    return checkboxes;
  }

  private outputFilters(filters: []) {
    const filterSections = [];

    for (const prop in filters) {
      if (filters.hasOwnProperty(prop)) {
        const params = filters[prop];
        filterSections.push(this.getFilterSection(params));
      }
    }

    return filterSections;
  }

  private outputRange(name: string, options: { min: string, max: string }) {
    return (
      <div>
        From $
        <input 
          type="text" 
          className="products-filter-property products-filter-price-from"
          name={`{name}[]`}
          value={options.min} 
          onChange={this.handlePropertyPriceRange}
          size={4} /> 
        to $
        <input 
          type="text" 
          className="products-filter-property products-filter-price-to"
          name={`{name}[]`}
          value={options.max}
          onChange={this.handlePropertyPriceRange}
          size={4} />
      </div>
    )
  }

  private handlePropertyCheck(e: any) {
    const check = e.target;
    const opts = Object.assign({}, this.options);

    if (check.checked) {
      if ( !opts[check.name] ) {
        opts[check.name] = [];
      }

      opts[check.name].push(check.value);
    } else {
      if ( opts[check.name] ) {
        const i = opts[check.name].indexOf(check.value);
        opts[check.name].splice(i, 1);
      }
    }

    this.options = opts;

    this.props.passFilters(opts);
  }

  private handlePropertyPriceRange() {
    const from = (document.querySelector('.products-filter-price-from') as HTMLInputElement);
    const to = (document.querySelector('.products-filter-price-to') as HTMLInputElement);

    if (to && to.value !== '') {
      this.options.priceTo = to.value;
      this.setState({ priceTo: to.value });
    }

    if (from && from.value !== '') {
      this.options.priceFrom = from.value;
      this.setState({ priceFrom: from.value });
    }

    this.props.passFilters(this.options);
  }
}

function mapStateToProps(state: any) {
  return {
    filters: state.filters.data,
    // ...state.products,
  }
}

function mapDispatchToProps(dispatch: any) {
  return {
    fetchFiltersByCategory: (filterCategory: string) => dispatch(fetchFiltersByCategory(filterCategory)),
    passFilters: (filterOptions: {}) => dispatch(passFilters(filterOptions)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(CatalogFilter));
