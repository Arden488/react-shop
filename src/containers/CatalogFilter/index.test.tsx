import { mount, shallow } from 'enzyme';
import * as React from 'react';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import CatalogFilter from './index';

const mockStore = configureStore([thunk]);
let store: any;
const initialState = {
  filters: { data: {
    type: {
      field_type: "checkbox",
      name: 'type',
      title: 'Type',
      variants: ["iPad 3", "iPad 4", "iPad 5", "iPad 6"]
    }
  } },
  products: { data: [] },
  productsFilterOptions: { data: {} },
};
let component: React.ReactNode = null;

beforeEach(() => {
  store = mockStore(initialState);
  component = mount(<Provider store={store}><CatalogFilter /></Provider>);
});

it('renders correctly', () => {
  expect(component).toMatchSnapshot();
});
