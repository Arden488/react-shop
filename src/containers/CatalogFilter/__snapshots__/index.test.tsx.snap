// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatch filter action 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "FILTER_PASSED",
  },
]
`;

exports[`renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "abcdef",
          "pathname": "/catalog/ipad",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "abcdef",
              "pathname": "/catalog/ipad",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "abcdef",
            "pathname": "/catalog/ipad",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(withRouter(CatalogFilter))>
        <withRouter(CatalogFilter)
          fetchFiltersByCategory={[Function]}
          filters={
            Object {
              "price_range": Object {
                "field_type": "range",
                "name": "price",
                "options": Object {
                  "max": "1000",
                  "min": "300",
                },
                "title": "Price range",
              },
              "type": Object {
                "field_type": "checkbox",
                "name": "type",
                "title": "Type",
                "variants": Array [
                  "iPad 3",
                  "iPad 4",
                  "iPad 5",
                  "iPad 6",
                ],
              },
            }
          }
          passFilters={[Function]}
        >
          <Route>
            <CatalogFilter
              fetchFiltersByCategory={[Function]}
              filters={
                Object {
                  "price_range": Object {
                    "field_type": "range",
                    "name": "price",
                    "options": Object {
                      "max": "1000",
                      "min": "300",
                    },
                    "title": "Price range",
                  },
                  "type": Object {
                    "field_type": "checkbox",
                    "name": "type",
                    "title": "Type",
                    "variants": Array [
                      "iPad 3",
                      "iPad 4",
                      "iPad 5",
                      "iPad 6",
                    ],
                  },
                }
              }
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "canGo": [Function],
                  "createHref": [Function],
                  "entries": Array [
                    Object {
                      "hash": "",
                      "key": "abcdef",
                      "pathname": "/catalog/ipad",
                      "search": "",
                    },
                  ],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "index": 0,
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "abcdef",
                    "pathname": "/catalog/ipad",
                    "search": "",
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "key": "abcdef",
                  "pathname": "/catalog/ipad",
                  "search": "",
                }
              }
              match={
                Object {
                  "isExact": false,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
              passFilters={[Function]}
            >
              <div>
                <div
                  className="catalog-filter__section"
                  key="price"
                >
                  <div>
                    Price range
                    :
                  </div>
                  <div>
                    <div>
                      From $
                      <input
                        className="products-filter-property products-filter-price-from"
                        name="{name}[]"
                        onChange={[Function]}
                        size={4}
                        type="text"
                        value="300"
                      />
                      to $
                      <input
                        className="products-filter-property products-filter-price-to"
                        name="{name}[]"
                        onChange={[Function]}
                        size={4}
                        type="text"
                        value="1000"
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="catalog-filter__section"
                  key="type"
                >
                  <div>
                    Type
                    :
                  </div>
                  <div>
                    <div
                      key="iPad 3"
                    >
                      <input
                        className="products-filter-property"
                        name="type"
                        onChange={[Function]}
                        type="checkbox"
                        value="iPad 3"
                      />
                      iPad 3
                    </div>
                    <div
                      key="iPad 4"
                    >
                      <input
                        className="products-filter-property"
                        name="type"
                        onChange={[Function]}
                        type="checkbox"
                        value="iPad 4"
                      />
                      iPad 4
                    </div>
                    <div
                      key="iPad 5"
                    >
                      <input
                        className="products-filter-property"
                        name="type"
                        onChange={[Function]}
                        type="checkbox"
                        value="iPad 5"
                      />
                      iPad 5
                    </div>
                    <div
                      key="iPad 6"
                    >
                      <input
                        className="products-filter-property"
                        name="type"
                        onChange={[Function]}
                        type="checkbox"
                        value="iPad 6"
                      />
                      iPad 6
                    </div>
                  </div>
                </div>
              </div>
            </CatalogFilter>
          </Route>
        </withRouter(CatalogFilter)>
      </Connect(withRouter(CatalogFilter))>
    </Router>
  </MemoryRouter>
</Provider>
`;
